cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/handlers/OpenScienceFrameworkAuthenticationHandler.java

```diff
@@ -138,22 +139,28 @@ protected final HandlerResult authenticateInternal(final OpenScienceFrameworkCre
         if (user == null) {
             throw new AccountNotFoundException(username + " not found with query");
         }
-
-        Boolean validPassphrase = Boolean.FALSE;
         final String userStatus = verifyUserStatus(user);
 
-        if (credential.isRemotePrincipal()) {
-            // verified through remote principals
-            validPassphrase = Boolean.TRUE;
-        } else if (verificationKey != null && verificationKey.equals(user.getVerificationKey())) {
-            // verified by verification key
-            validPassphrase = Boolean.TRUE;
-        } else if (plainTextPassword != null && verifyPassword(plainTextPassword, user.getPassword())) {
-            // verified by password
-            validPassphrase = Boolean.TRUE;
-        }
-        if (!validPassphrase) {
-            throw new FailedLoginException(username + ": invalid remote authentication, verification key or password");
+        // Verify the user's credential
```

Comment:
Rewrite credential verification in the authentication handler:
* Each condition are more explicit
* Exceptions are thrown close to where failures happen
* Add a dedicated exception for invalid verification key